File: go/cedarcork/cedarcork.go
package cedarcork

import (
	"crypto/rand"
	"errors"
	"fmt"
	"time"

	corks "github.com/celest-dev/corks/go"
	"github.com/celest-dev/corks/go/cedar"
	cedarexpr "github.com/celest-dev/corks/go/cedar/expr"
	cedarv4 "github.com/celest-dev/corks/go/proto/cedar/v4"
	corksv1 "github.com/celest-dev/corks/go/proto/celest/corks/v1"
	"google.golang.org/protobuf/types/known/anypb"
)

const (
	cedarNamespace              = "celest.cedar"
	cedarPredicate              = "expr"
	defaultCaveatVersion uint32 = 1
)

type Cork struct {
	*corks.Cork
}

func (c *Cork) Issuer() *cedar.EntityUid {
	if c == nil {
		return nil
	}
	issuer := c.Cork.Issuer()
	if issuer == nil {
		return nil
	}
	entityId := new(cedarv4.EntityUid)
	err := issuer.UnmarshalTo(entityId)
	if err != nil {
		return nil
	}
	return (*cedar.EntityUid)(entityId)
}

func (c *Cork) Bearer() *cedar.EntityUid {
	if c == nil {
		return nil
	}
	bearer := c.Cork.Bearer()
	if bearer == nil {
		return nil
	}
	entityId := new(cedarv4.EntityUid)
	err := bearer.UnmarshalTo(entityId)
	if err != nil {
		return nil
	}
	return (*cedar.EntityUid)(entityId)
}

func (c *Cork) Audience() *cedar.EntityUid {
	if c == nil {
		return nil
	}
	audience := c.Cork.Audience()
	if audience == nil {
		return nil
	}
	entityId := new(cedarv4.EntityUid)
	err := audience.UnmarshalTo(entityId)
	if err != nil {
		return nil
	}
	return (*cedar.EntityUid)(entityId)
}

func (c *Cork) Claims() *cedarv4.Entity {
	if c == nil {
		return nil
	}
	claims := c.Cork.Claims()
	if claims == nil {
		return nil
	}
	entity := new(cedarv4.Entity)
	err := claims.UnmarshalTo(entity)
	if err != nil {
		return nil
	}
	return entity
}

func (c *Cork) Caveats() []*cedarv4.Expr {
	if c == nil {
		return nil
	}
	var expressions []*cedarv4.Expr
	for _, caveat := range c.Cork.Caveats() {
		first := caveat.GetFirstParty()
		if first == nil || first.Namespace != cedarNamespace || first.Predicate != cedarPredicate {
			continue
		}
		if first.Payload == nil {
			continue
		}
		expression := new(cedarv4.Expr)
		if err := first.Payload.UnmarshalTo(expression); err != nil {
			return nil
		}
		expressions = append(expressions, expression)
	}
	return expressions
}

// Rebuild returns a new builder with the cork's data.
func (c *Cork) Rebuild() *builder {
	b := &builder{Builder: c.Cork.Rebuild()}
	return b
}

type builder struct {
	*corks.Builder
	errors []error
}

// NewBuilder creates a new cork builder with the given key identifier.
func NewBuilder(keyID []byte) *builder {
	return &builder{Builder: corks.NewBuilder(keyID)}
}

// Issuer sets the issuer of the cork.
func (b *builder) Issuer(issuer *cedar.EntityUid) *builder {
	b.Builder.Issuer(issuer.Proto())
	return b
}

// Bearer sets the bearer of the cork.
func (b *builder) Bearer(bearer *cedar.EntityUid) *builder {
	b.Builder.Bearer(bearer.Proto())
	return b
}

// Audience sets the audience of the cork.
func (b *builder) Audience(audience *cedar.EntityUid) *builder {
	b.Builder.Audience(audience.Proto())
	return b
}

// Claims sets the claims of the cork.
func (b *builder) Claims(claims *cedar.Entity) *builder {
	b.Builder.Claims(claims.Proto())
	return b
}

// Caveat adds a caveat to the cork.
func (b *builder) Caveat(caveat *cedarexpr.Expr) *builder {
	if caveat == nil {
		b.errors = append(b.errors, fmt.Errorf("%w: caveat is nil", corks.ErrInvalidCork))
		return b
	}
	payload, err := anypb.New(caveat.Proto())
	if err != nil {
		b.errors = append(b.errors, err)
		return b
	}
	id := make([]byte, 16)
	if _, err := rand.Read(id); err != nil {
		b.errors = append(b.errors, err)
		return b
	}
	protoCaveat := &corksv1.Caveat{
		CaveatVersion: defaultCaveatVersion,
		CaveatId:      id,
		Body: &corksv1.Caveat_FirstParty{
			FirstParty: &corksv1.FirstPartyCaveat{
				Namespace: cedarNamespace,
				Predicate: cedarPredicate,
				Payload:   payload,
			},
		},
	}
	b.Builder.Caveat(protoCaveat)
	return b
}

// NotAfter sets the expiry on the underlying cork builder.
func (b *builder) NotAfter(notAfter time.Time) *builder {
	b.Builder.NotAfter(notAfter)
	return b
}

// IssuedAt allows tests to override the issuance timestamp.
func (b *builder) IssuedAt(issuedAt time.Time) *builder {
	b.Builder.IssuedAt(issuedAt)
	return b
}

// Validate validates a cork before signing.
//
// A cork is valid if all required fields are present and all fields are valid.
func (b *builder) Validate() error {
	if err := b.Builder.Validate(); err != nil {
		return err
	}
	if len(b.errors) > 0 {
		return fmt.Errorf("%w: %v", corks.ErrInvalidCork, errors.Join(b.errors...))
	}
	return nil
}

// Build builds and validates the cork for signing.
func (b *builder) Build() (*Cork, error) {
	cork, err := b.Builder.Build()
	if err != nil {
		return nil, err
	}
	return &Cork{cork}, nil
}

File: dart/lib/src/cedar_cork.dart
import 'dart:typed_data';

import 'package:cedar/ast.dart' as cedar;
import 'package:cedar/cedar.dart' as cedar_model;
import 'package:cedar/src/proto/cedar/v4/entity.pb.dart' as cedar_entity;
import 'package:cedar/src/proto/cedar/v4/entity_uid.pb.dart' as cedar_uid;
import 'package:cedar/src/proto/cedar/v4/expr.pb.dart' as cedar_expr;
import 'package:corks_cedar/src/crypto.dart';
import 'package:corks_cedar/src/proto.dart';
import 'package:meta/meta.dart';

import 'cork.dart';
import 'proto/celest/corks/v1/cork.pb.dart' as corksv1;
import 'proto/google/protobuf/any.pb.dart' as anypb;

/// Cedar-specialized view over a [Cork] credential.
///
/// This extension type exposes Cedar domain objects instead of raw protobuf
/// payloads so applications can work with strongly typed entities, claims, and
/// caveat expressions.
extension type CedarCork(Cork _cork) implements Cork {
  /// Parses a base64-encoded [CedarCork].
  factory CedarCork.parse(String token) => CedarCork(Cork.parse(token));

  /// Decodes a binary-encoded [CedarCork].
  factory CedarCork.decode(Uint8List bytes) => CedarCork(Cork.decode(bytes));

  /// Creates a [CedarCork] from its protocol buffer representation.
  factory CedarCork.fromProto(corksv1.Cork proto) =>
      CedarCork(Cork.fromProto(proto));

  /// Creates a [CedarCork] from its JSON representation.
  factory CedarCork.fromJson(Map<String, Object?> json) =>
      CedarCork(Cork.fromJson(json));

  /// Creates a new [CedarCorkBuilder].
  static CedarCorkBuilder builder(Uint8List keyId) =>
      CedarCorkBuilder._(Cork.builder(keyId));

  static const int caveatVersion = 1;
  static const String caveatNamespace = 'celest.cedar';
  static const String caveatPredicate = 'expr';

  /// Cedar entity that issued this cork, if present.
  @redeclare
  cedar.EntityUid? get issuer => _decodeEntityUid(_cork.issuer);

  /// Cedar entity representing the bearer, if present.
  @redeclare
  cedar.EntityUid? get bearer => _decodeEntityUid(_cork.bearer);

  /// Cedar entity representing the audience, if present.
  @redeclare
  cedar.EntityUid? get audience => _decodeEntityUid(_cork.audience);

  /// Structured Cedar claims packed into the cork, if present.
  @redeclare
  cedar.Entity? get claims => _decodeEntity(_cork.claims);

  /// Cedar caveat expressions embedded as first-party caveats.
  @redeclare
  List<cedar.Expr> get caveats {
    final expressions = <cedar.Expr>[];
    for (final caveat in _cork.caveats) {
      if (!caveat.hasFirstParty()) {
        continue;
      }
      final first = caveat.firstParty;
      if (first.namespace != caveatNamespace ||
          first.predicate != caveatPredicate) {
        continue;
      }
      if (!first.hasPayload()) {
        continue;
      }
      final expr = _decodeExpr(first.payload);
      if (expr != null) {
        expressions.add(expr);
      }
    }
    return expressions;
  }

  CedarCorkBuilder rebuild() => CedarCorkBuilder._(_cork.rebuild());
}

/// Builder for cork credentials that embed Cedar payloads.
extension type CedarCorkBuilder._(CorkBuilder _builder) implements CorkBuilder {
  /// Creates a Cedar-aware builder for the given [keyId].
  factory CedarCorkBuilder.forKey(Uint8List keyId) =>
      CedarCorkBuilder._(CorkBuilder(keyId));

  @redeclare
  set issuer(cedar_model.EntityUid issuer) {
    _builder.issuer = issuer.toProto();
  }

  @redeclare
  set bearer(cedar_model.EntityUid bearer) {
    _builder.bearer = bearer.toProto();
  }

  @redeclare
  set audience(cedar_model.EntityUid? audience) {
    _builder.audience = audience?.toProto();
  }

  @redeclare
  set claims(cedar_model.Entity? claims) {
    _builder.claims = claims?.toProto();
  }

  @redeclare
  void addCaveat(cedar.Expr expression) {
    final payload = expression.toProto().packIntoAny();
    final id = secureRandomBytes(16);
    final caveat = corksv1.Caveat(
      caveatVersion: CedarCork.caveatVersion,
      caveatId: id,
      firstParty: corksv1.FirstPartyCaveat(
        namespace: CedarCork.caveatNamespace,
        predicate: CedarCork.caveatPredicate,
        payload: payload,
      ),
    );
    _builder.addCaveat(caveat);
  }
}

cedar_model.EntityUid? _decodeEntityUid(anypb.Any? any) {
  if (any == null || any.value.isEmpty) {
    return null;
  }
  final proto = cedar_uid.EntityUid()..mergeFromBuffer(any.value);
  return cedar_model.EntityUid.fromProto(proto);
}

cedar_model.Entity? _decodeEntity(anypb.Any? any) {
  if (any == null || any.value.isEmpty) {
    return null;
  }
  final proto = cedar_entity.Entity()..mergeFromBuffer(any.value);
  return cedar_model.Entity.fromProto(proto);
}

cedar.Expr? _decodeExpr(anypb.Any payload) {
  if (payload.value.isEmpty) {
    return null;
  }
  final proto = cedar_expr.Expr()..mergeFromBuffer(payload.value);
  return cedar.Expr.fromProto(proto);
}
